### ISSUE TITLE
Corruption on returning composite ops from functions

### ISSUE THREAD

**Issue by feltech** [romeric/Fastor#95]
A function such as the following (with `b` and `I` being `Tensor`s), which returns a composite op, cannot safely be assigned to a `Tensor` because the branches of the op tree go out of scope and are destroyed after the function returns.
```cpp
auto const sigma = [](
	Scalar const J, auto const & b,
	Scalar const lambda, Scalar const mu) {
	return (mu / J) * (b - I) + (lambda / J) * log(J) * I;
};
```
It seems the construction of the op tree relies on `const` references (`expression_t`) to the branches, as opposed to moves or copies.  So outside the function, those `const` references are now (probably) invalid and the op segfaults when evaluated.

I noticed the `FASTOR_COPY_EXPR` flag is available, and seems to solve the problem.  But of course that's a global flag and there's presumably good performance reasons why we should use references whenever possible.

A better solution for the above case in particular is perhaps changing `expression_t` from 
```
conditional_t_<is_arithmetic_v_<T>, const T, const T&>
```
to
```
conditional_t_<(is_expression_v<T> || is_arithmetic_v_<T>), const T, const T&>
```
so that ops are copied but tensors at the leafs remain references.  This of course doesn't work if the function above was returning an op that relied on a temporary tensor created inside the function. In that case we would want to copy the tensor rather than attempting to retain a reference to it.  I'm not sure how to detect that case.

We can additionally add operator overloads that accept r-value references in `binary_arithmetic_ops.h`.  I tried that and added some logging (since debugger is useless for macros) and they do get hit.  Presumably this could ensure op branches are moved into their parent rather than copied.  Though I'm not sure it's worth it - there isn't much data and the compiler can probably elide perfectly well on it's own.

Annoyingly, trying to reproduce this problem in a minimal example isn't so easy (on gcc 7.5), presumably because the memory is still safe to access for some limited time, despite it going out of scope (I've seen this "feature" of gcc before).

**Comment by romeric** [comment]
@feltech You have discovered the most fundamental bug in Fastor's expression template engine. I am aware of this issue and I agree this bug is pretty hard to reproduce and I reckon that is partly because of the static storage of Tensors. If Fastor's tensors were heap allocated this bug would show-up everywhere. Perhaps we can test this with much bigger tensors for debugging? or I believe Valgrind would be able to detect access to destroyed stack frames.

Yes, `FASTOR_COPY_EXPR` unfortunately does hit a couple of benchmarks pretty badly (at least when I tested it a long time ago - circa 2016). It also almost defeats the purpose of building a tree but the main reason it is there is because of this issue. 

Now, I was going to make the exact same change that you mentioned i.e.

~~~c++
conditional_t_<(is_expression_v<T> || is_arithmetic_v_<T>), const T, const T&>
~~~

prior to `0.6` release but somehow forgot about it and it didn't make it to that release. That will address most of the cases with operating chaining. Please go ahead and submit a PR with this change, you already know the internals of Fastor more than I do!

Now regarding the case where a tree containing a local tensor is returned to the outer scope, I have been at a loss myself. We have had one or two issues reported here in the past concerning this issue (I will ref them if I find them) and my advice so far has simply been "don't use `auto` as a return type when you are returning an expression". Which is pretty dumb but hey every math expression template library has had one or two fundamental bug(s) that has taken years for it to be resolved (look up "Eigen bug 99" for instance if you don't know of it already). 

Regardless this remains an issue and hopefully we can resolve it sooner than later.

**Comment by romeric** [comment]
FYI, I am able to easily reproduce the bug on macOS with Apple Clang and on Windows 10 with Visual Studio 2019 with this 

~~~c++
using Scalar = double;
constexpr size_t M = 30;
Tensor<Scalar,M,M> I; I.eye();
Tensor<Scalar,M,M> b; b.iota(2);
Scalar lambda = 100.;
Scalar mu = 50.;
Scalar J = 0.98;

auto const sigma = [&I](
    Scalar const J, auto const & b,
    Scalar const lambda, Scalar const mu) {
    return (mu / J) * (b - I) + (lambda / J) * log(J) * I;
};

auto s = sigma(J,b,lambda,mu);
print(s);
~~~

~~~
(base) roman@Romans-MBP % ./main
segmentation fault  ./main
~~~

If I copy the expression using your suggested change, the bug understandably disappears with both clang and MSVC. `GCC-9.2` is fine with and without the change.

It should come as a consolation that the following examples

~~~c++
constexpr size_t mm = 20;
auto foo() {
    Tensor<double,mm,mm> a(2);
    Tensor<double,mm,mm> b(2);
    return a + b;
}

auto c = foo();
Tensor<double,mm,mm> d = a;
~~~

and the Eigen equivalent

~~~c++
constexpr size_t mm = 20;
auto foo() {
    Eigen::Matrix<double,mm,mm> a; a.setConstant(2);
    Eigen::Matrix<double,mm,mm> b; b.setConstant(2);
    return a + b;
}

auto c = foo();
Eigen::Matrix d = a;
~~~

are both buggy in the same sense. I also tested the `xtensor` package and same story there. So perhaps we should question if the C++14 `auto` return keyword should really be used in conjunction with expressions.

**Comment by feltech** [comment]
Ah very interesting.  I am actually migrating code from Eigen to Fastor and didn't experience this problem with Eigen, though that was probably just luck.

Like I hinted, I've been caught out before with the annoying gcc "feature" of allowing invalid memory access in previous work, which only came to light when QA tested the MSVC build.  I'll have to investigate linters / compiler switches / sanitizers, see if anything can pick it up (any suggestions?).

In general it's surely useful to be able to reuse expressions.  The specific case that motivated this bug report was unit tests, where I want to test pieces of a larger expression, and combine those pieces in the final function, without repeating code.

I guess it may just be reasonable to warn people that here-be-dragons with auto return type.  

The only workaround I can think of off the top of my head is to use macros instead of functions, which I'd rather avoid.  Do you have any suggestions?

**Comment by romeric** [comment]
I don't think the issue of allowing access to invalid frames is endemic to GCC only. MSVC and Intel compilers seem to be really good in this regard though. If you find GCC tripping you up often on this issue, use the memory sanitiser. As I mentioned earlier this problem does not show up partly because of static storage of tensors and we need a way to change that to heap allocation (just for internal debugging). Memory sanitisers are exactly that in that they change/mock stack frames to heap.

Here is an example GCC's AddressSanitizer on the `foo` example I posted above

~~~ shell
$ g++ main.cpp -o main -fsanitize=address
env ASAN_OPTIONS='detect_stack_use_after_return=1' ./main
~~~

~~~
=================================================================
==91684==ERROR: AddressSanitizer: stack-use-after-return on address 0x000109895d80 at pc 0x00010558ecd7 bp 0x7ffeea66edc0 sp 0x7ffeea66edb8
READ of size 16 at 0x000109895d80 thread T0
    #0 0x10558ecd6 in main (main:x86_64+0x100002cd6)
    #1 0x7fff73b9dcc8 in start (libdyld.dylib:x86_64+0x1acc8)

Address 0x000109895d80 is located in stack of thread T0 at offset 3456 in frame
    #0 0x10558d3f6 in foo() (main:x86_64+0x1000013f6)

  This frame has 5 object(s):
    [32, 48) '_vec'
    [64, 80) '_vec'
    [96, 112) '<unknown>'
    [128, 3328) 'a' (line 160)
    [3456, 6656) 'b' (line 161) <== Memory access at offset 3456 is inside this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-return (main:x86_64+0x100002cd6) in main
Shadow bytes around the buggy address:
  0x100021312b60: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312b70: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312b80: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312b90: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312ba0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
=>0x100021312bb0:[f5]f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312bc0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312bd0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312be0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312bf0: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x100021312c00: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==91684==ABORTING
zsh: abort      env ASAN_OPTIONS='detect_stack_use_after_return=1' ./main
~~~

and yes Eigen, Fastor and all other expression template libraries have the same issue (as far as returning a local tensor as part of an expression is concerned). In fact Eigen's philosophy is the same as what we have discussed so far.

~~~c++
if_t<is_op_v<T>,const T&, const T>
~~~

I don't think any of the strategies that I have thought of would be worth implementing and I don't think I have a good one either. I wouldn't use macros for this. Also, I removed all move semantics from Fastor over a year or so ago simply because in the last 3 years of using Fastor in my personal projects I haven't come across a case where the compiler didn't elide the copy.

**Comment by romeric** [comment]
Pushed the change here [5f9d45e] and updated the [Wiki page](https://github.com/romeric/Fastor/wiki/Pitfalls-&-gotchas) on the use of `auto`.

For reference here are the equivalent pages from XTensor and Eigen projects on the same topic

https://xtensor.readthedocs.io/en/latest/pitfall.html#intermediate-result

https://eigen.tuxfamily.org/dox/TopicPitfalls.html

On the topic of re-using an expression, while I agree that it is extremely useful you also have to consider that the more an expression is being re-used the more the chances of potentially traversing the expression tree. This will certainly keep the memory footprint low but at the cost of doubling/trippling/quadrippling/... the number of floating point operations (unless the expression is evaluted only once). In your function above I believe you are computing some sort of a stress tensor and you have a `log` which is particularly expensive and you don't want that to be evaluated twice unless this is part of a bigger expression and only that expression gets evaluated at the end (and not the stress tensor itself).

**Comment by feltech** [comment]
Thanks for adding that change in, and the advice on using the sanitizer, much appreciated. You make all good points.  I'll close this issue now.  Thanks for the detailed treatment of this issue!
